src/gotcha_dl.c

```diff
@@ -5,7 +5,124 @@
+ * @param info, information about headers
+ * @param size, size of headers
+ * @param data, data from caller
+ * @return
+ */
+int lib_header_callback(struct dl_phdr_info * info, size_t size, void * data) {
+    struct Addrs* addrs = data;
+    ElfW(Addr) end_addr = 0;
+    for (int i=0;i<info->dlpi_phnum;++i) {
+        ElfW(Addr) size = info->dlpi_phdr[i].p_vaddr + info->dlpi_phdr[i].p_memsz;
+        if (end_addr < size) end_addr = size;
+    }
+    ElfW(Addr) base_addr = info->dlpi_addr;
+    ElfW(Addr) end_phdr = base_addr + end_addr;
+    if (base_addr <= addrs->lookup_addr && end_phdr >= addrs->lookup_addr)
+        addrs->library_laddr = base_addr;
+    return 0;
+}
+/**
+ * This function looks for all symbols defined in rela or rel.
+ *
+ * @param name, symbol name
+ * @param lmap, library where we search for symbol.
+ * @return symbol pointer
+ */
+int dl_lookup_symbol_x(const char *name, struct link_map *lmap, void** symbol) {
+    INIT_DYNAMIC(lmap)
+    ElfW(Addr) offset = lmap->l_addr;
+    (void) offset;
+    if (is_rela) {
+        rela = (ElfW(Rela) *) jmprel;
+        for (i = 0; i < rel_count; i++) {
+          ElfW(Addr) offset = rela[i].r_offset;
+          unsigned long symidx = R_SYM(rela[i].r_info);
+          ElfW(Sym) *sym = symtab + symidx;
+          char *symname = strtab + sym->st_name;
+          if (strcmp(symname, name) == 0) {
+            symbol = (void **) (lmap->l_addr + offset);
+            return 0;
+          }
+       }
+    } else {
+        rel = (ElfW(Rel) *) jmprel;
+        for (i = 0; i < rel_count; i++) {
+            ElfW(Addr) offset = rel[i].r_offset;
+            unsigned long symidx = R_SYM(rel[i].r_info);
+            ElfW(Sym) *sym = symtab + symidx;
+            char *symname = strtab + sym->st_name;
+            if (strcmp(symname, name) == 0) {
+                symbol = (void **) (lmap->l_addr + offset);
+                return 0;
+            }
+        }
+    }
+    return -1;
+}
+
+/**
+ * Implement the logic of _dl_sym
+ * 1. find the caller library using the program headers.
+ * 2. find the first library which has the symbol for RTLD_DEFAULT
+ * 3. find the second library which has the symbol for RTLD_NEXT
+ * @param handle, handle for dl operation
+ * @param name, name of the symbol
+ * @param who, the virtual address of the caller
+ * @return symbol pointer
+ */
+
+static void * _dl_sym(void *handle, const char *name, void *who) {
+    ElfW(Addr) caller = (ElfW(Addr)) who;
+    /* Iterative over the library headers and find the caller
+     * the address of the caller is set in addrs->library_laddr
+     **/
+    struct Addrs addrs;
+    addrs.lookup_addr = caller;
+    dl_iterate_phdr(lib_header_callback, &addrs);
+
+    struct link_map *lib_iter;
+    int found_caller = 0;
+    int found_sym = 0;
+    for (lib_iter = _r_debug.r_map; lib_iter != 0; lib_iter = lib_iter->l_next) {
+        /* find the library of the caller */
+        if (lib_iter->l_addr == addrs.library_laddr)
```

Comment:
the ```dl_phdr_info``` does not contain the lmap or give any pointer to lmap. So I used the name to match the corresponding lmap.

Reply:
Hmm. You're right on the link_map pointer not being there. But name won't quite be unique either. On some older GLIBC versions both the VDSO and the executable have blank name strings. 

I think you could use both name and load-address for uniqueness. I don't recollect any system where they wouldn't be unique when combined.

Reply:
Ok i added load_address which i am thinking is the base address which is info->dlpi_addr. As I understand that this can be 0 in which case we check the name.